das/aggregator.go

```diff
@@ -0,0 +1,218 @@
+			if blobHash == requestedCert.DataHash {
+				blobChan <- blob
+			} else {
+				errorChan <- fmt.Errorf("DAS (mask %X) returned data that doesn't match requested hash!", d.signerMask)
+			}
+		}(subCtx, d)
+	}
+
+	errorCount := 0
+	var errorCollection []error
+	for errorCount < len(servicesThatSignedCert) {
+		select {
+		case blob := <-blobChan:
+			return blob, nil
+		case err = <-errorChan:
+			errorCollection = append(errorCollection, err)
+			log.Warn("Couldn't retrieve message from DAS", "err", err)
+			errorCount++
+		case <-ctx.Done():
+			break
+		}
+	}
+
+	return nil, fmt.Errorf("Data wasn't able to be retrieved from any DAS: %v", errorCollection)
+}
+
+type storeResponse struct {
+	cert *arbstate.DataAvailabilityCertificate
+	err  error
+
+	details serviceDetails
+}
+
+// Store calls Store on each backend DAS in parallel and collects responses; if
+// there were enough (at least K) responses then it aggregates the signatures and
+// signerMasks from each DAS together to put in the DataAvailabilityCertificate
+// that it returns; if it gets enough errors that K successes is impossible, then
+// it stops early and returns an error, and if it gets not enough successful
+// responses by the time its context is canceled (eg via DeadlineWrapper) then it
+// also returns an error.
+func (a *Aggregator) Store(ctx context.Context, message []byte, timeout uint64) (*arbstate.DataAvailabilityCertificate, error) {
+	var aggSignersMask uint64
+	var pubKeys []blsSignatures.PublicKey
+	var sigs []blsSignatures.Signature
+	var aggCert arbstate.DataAvailabilityCertificate
+
+	var initialStoreSucceeded bool
+	storeFailures := 0
+	if timeout == CALLEE_PICKS_TIMEOUT {
+		timeout = uint64(time.Now().Add(a.config.retentionPeriod).Unix())
+	}
+
+	responses := make(chan storeResponse)
+	subCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for _, d := range a.services {
+		go func(ctx context.Context, d serviceDetails) {
+			cert, err := d.service.Store(ctx, message, timeout)
+			responses <- storeResponse{cert, err, d}
+		}(subCtx, d)
+	}
+
+	successfullyStoredCount := 0
+	for i := 0; i < len(a.services); i++ {
+		select {
+		case <-ctx.Done():
+			if successfullyStoredCount < a.requiredServicesForStore {
+				return nil, errors.New("Terminated das.Aggregator.Store() with requests outstanding before enough responses received")
+			} else {
+				break
+			}
+		case r := <-responses:
+			if r.err != nil {
+				storeFailures++
+				log.Warn("Failed to store message to DAS", "err", r.err)
+				if storeFailures <= a.maxAllowedServiceStoreFailures {
+					continue
+				} else {
+					return nil, fmt.Errorf("Aggregator failed to store message to at least %d out of %d DASes (assuming %d are honest)", a.requiredServicesForStore, len(a.services), a.config.assumedHonest)
+				}
+			}
+			verified, err := blsSignatures.VerifySignature(r.cert.Sig, serializeSignableFields(*r.cert), r.details.pubKey)
+			if err != nil {
```

Comment:
We don't want to fail the `Store` because of one failed signature verification.  (Otherwise a malicious DAS could fail the aggregate Store operation by submitting a bad signature. Instead we should just treat the response as an error.

Reply:
Thanks, I hadn't noticed this because I hadn't made the Store test do data corruption, so I fixed that. I made it so all the errors encountered when collecting responses work like this now.